### DISCUSSION TITLE
Could we provide a more flexible `get<>` and `at<>` API?

### DISCUSSION THREAD

**Discussion by sighingnow** [nlohmann/json#2543] [Ideas]
#### What is the issue you have?

(I didn't see a "feature request" category when opening a new issue thus I post this ticket as a "bug report")

I'm wondering could we provides a more flexible `get<>` (`get` with key) or `at<>` (`at` with type parameter) API like:

```cpp
template <typename ValueType>
ValueType get(const key_type &key) {
   return this->operator[](key).get<ValueType>();
}

template <typename ValueType>
ValueType at(const key_type &key) {
   return this->at(key).get<ValueType>();
}
```

I have notice we already has

+ `get` with type parameter but without key
+ `at` and `[]` with key but no type parameter
+ `value` with key but requires a typed default value.

I think the new API above would provide great convenience.

#### Please describe the steps to reproduce the issue.

None.

#### Can you provide a small but working code example?

None.

#### What is the expected behavior?

As the description of this issue.

#### And what is the actual behavior instead?

No such API.

#### Which compiler and operating system are you using?

None.

**Comment by sighingnow** [comment]
If such API is acceptable, I could prepare a pull request for that.

**Comment by sighingnow** [comment]
Also for `get_ref<>`.

**Comment by nickaein** [comment]
Can you provide some code snippets that demonstrate the usage of this API?

BTW, there is a "Ideas" category at [Discussions](https://github.com/nlohmann/json/discussions/) which seems to be meant for submitting proposals.

**Reply by sighingnow** [reply]
For example we have a json:

```
{
    "name": "foo",
    "age": 10,
}
```

And we want to get the age, before this proposal we need `data["age"].get<int>()`, after this proposal we just need `data.get<int>("age")`. The later is more concise and self-explaining IMO.

**Reply by sighingnow** [reply]
See also data accessor in boost::ptree: https://www.boost.org/doc/libs/1_65_1/doc/html/property_tree/accessing.html

**Reply by sighingnow** [reply]
@nickaein Any comment?

**Comment by gregmarr** [comment]
I don't see how the newer version is any more concise or self-explaining.  It doesn't remove the need for `json["keyname"]` nor the need for `json.get<value_type>()`.  It removes exactly two characters, which are punctuation.  It puts the type before the key being looked up, breaking the association between the key and the `json` object.  It is only valid for `json` variables that are objects, not those of any other type.

    json["keyname"].get<value_type>()
    json.get<value_type>("keyname")

It has the additional downside that it muddies the definition of `get`.  The function can currently mean "get the value of this `json` variable which is of the given type".  This would add an additional meaning of "check this `json` to see if this is an object.  If it is an object, check to see if it has a nested value under the key "keyname".  If it does, get the value of that nested object which is the given type.  If any of those new things are not true, then .... <insert fallback behavior(s) here>."

The current method is not something that is difficult or wordy to do.

**Reply by sighingnow** [reply]
I agree with the part of "breaking the association between the key and  the `json` object". However the implementation of such rountines doesn't need to be such complicate, as shown in the description of this discussion, it can be merely implemented as a chain of `[]` and `get`.

It doens't just remove two characters, for user interface, it saves a function call.

**Reply by gregmarr** [reply]
So I read your initial post again after writing the points below, and it seems that `value()` already does exactly what you want, except that it `requires a typed default value.`  That's not quite true, as you can explicitly specify the template type, as you are asking for, and then you can provide an untyped value. 

    data.value<int>("age", {});

or

    data.value<int>("age", 0);

and if the `, {}` or `, 0` at the end is too much, then a better proposal would be to add a default to the final `value()` parameter.

https://godbolt.org/z/7xznqn


Original preserved as I think it would be useful context:
It moves an `operator[]` (which is a function call in this case) from one place (user code) to another (nested inside get), so it's not truly saving it.

It's also increasing code complexity and the cognitive complexity of the API.  It is conflating the currently separated ideas of "get a nested json value from this json object/array" and "get the C++ value of this non-object/non-array json value".  These two ideas have different error case concerns, and the former has const vs non-const behavior differences.

Let's say that this function does get added, so there are now `get*<>` calls that also do `at()` in the same call.  There are four variants of `at()`, const and non-const, index and string.  There are currently 7 variants of the `get<>` calls, 13 when you include `get_ref<>`, `get_ptr<>`, and `get_to<>`.  To be able to cover everything that `json[key].get*<type>()` can do now as `json.get*<type>(key)` would require adding 52 more variants of these `get` functions (13*4).

Wouldn't it then make sense to be able to pass multiple key names to it, so we can collapse multiple `[]` calls?  If you do the same for taking "array of string" to support `json["foo"]["bar"]`, "array of integer" to support `json[0][1][2]`, and "json path", that's an additional 78 variants (13 `get_*` calls, const and non-const, 3 parameter types).  That also doesn't support `json["foo"][1]["bar"]` which is something you can do now.

**Reply by sighingnow** [reply]
Thanks for clearify the consideration behind current design of `get*` API. Now I can see the drawback of merging such two API into ones. For the proposal of "add a default to the final `value()`", I prefer adding a new method called `value_or_default<T>(key)`, for any type that statisfy `std::is_default_constructible<T>`.

Any comment ? @nickaein

**Reply by gregmarr** [reply]
I don't think an additional function is needed.  Adding `= {}` would make it work without that parameter for any default constructible type, and fail appropriately if it's not default constructible.  I'm not sure if this was previously considered in the design of this function.  @nlohmann would probably remember.

**Reply by nlohmann** [reply]
So you would propose chaning

```cpp
template<class ValueType>
ValueType value(const typename object_t::key_type& key,
                const ValueType& default_value) const;
```

to

```cpp
template<class ValueType>
ValueType value(const typename object_t::key_type& key,
                const ValueType& default_value = {}) const;

```

?

**Reply by sighingnow** [reply]
Agree with @gregmarr ,  I realized that it could fail appropriately if it isn't is_default_constructible and no default value.

Just add a default parameter `= {}` is fine for me.

**Reply by gregmarr** [reply]
I don't have a particular need for it, but if @sighingnow thinks it would make the function more useful, then yes.

**Reply by sighingnow** [reply]
Thank you @gregmarr, I will make a pull request for it.
